### DISCUSSION TITLE
getting "deprecated" warnings at startup, no text output in GUI

### DISCUSSION THREAD

**Discussion by jose97ct** [jas3333/GPT-Helper#5] [Q&A]
Hi, first of all thank you for making this available for everyone that wants to run chatgpt locally!!

The problem is when I execute `npm start`, the console output this message so quickly I haven't barely noticed it until now

```
user@server:~/GPT-Helper/client$ npm start

> GPT3Chat@0.1.0 start
> react-scripts start

(node:116690) [DEP_WEBPACK_DEV_SERVER_ON_AFTER_SETUP_MIDDLEWARE] DeprecationWarning: 'onAfterSetupMiddleware' option is deprecated. Please use the 'setupMiddlewares' option.
(node:116690) [DEP_WEBPACK_DEV_SERVER_ON_BEFORE_SETUP_MIDDLEWARE] DeprecationWarning: 'onBeforeSetupMiddleware' option is deprecated. Please use the 'setupMiddlewares' option.
Starting the development server...

Compiled successfully!

You can now view GPT3Chat in the browser.

  Local:            http://localhost:3000
  On Your Network:  http://192.168.1.105:3000

Note that the development build is not optimized.
To create a production build, use npm run build.

webpack compiled successfully
```

It's a freshly installed Ubuntu 22.01.1 LTS. I've just followed your instructions for the install process and created the file named ".env" as it is. I've tried both in VirtualBox and physical machine, and Firefox and Opera browsers.

I don't really know if that could be the reason I'm not getting any response when I type in the text box. The GUI is loading but it doesn't seem to be working properly for me. 
One last question is, do I need to do the `npm run build` thing?
Thanks in advance!

**Comment by jose97ct** [comment]
I've just tried again but in GitHub Codespaces, same result. Could it be a java version issue? Also I've noticed that my OpenAI key has an "-" on it, but I doubt it can cause a problem because I've tried with 3 more keys and "Last used" row in OpenAI webpage always says "Never"

**Comment by jose97ct** [comment]
I'm trying my best finding possible causes, this time I've found out these warnings during npm install.
Btw I installed npm via `aptitude install npm`

Console output:

```
usuario@server:~/GPT-Helper/client$ npm install
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: '@csstools/postcss-trigonometric-functions@1.0.2',
npm WARN EBADENGINE   required: { node: '^14 || >=16' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: '@jest/expect-utils@29.3.1',
npm WARN EBADENGINE   required: { node: '^14.15.0 || ^16.10.0 || >=18.0.0' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: '@jest/schemas@29.0.0',
npm WARN EBADENGINE   required: { node: '^14.15.0 || ^16.10.0 || >=18.0.0' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: '@jest/types@29.3.1',
npm WARN EBADENGINE   required: { node: '^14.15.0 || ^16.10.0 || >=18.0.0' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'pretty-format@29.3.1',
npm WARN EBADENGINE   required: { node: '^14.15.0 || ^16.10.0 || >=18.0.0' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'diff-sequences@29.3.1',
npm WARN EBADENGINE   required: { node: '^14.15.0 || ^16.10.0 || >=18.0.0' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'eslint-config-react-app@7.0.1',
npm WARN EBADENGINE   required: { node: '>=14.0.0' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'expect@29.3.1',
npm WARN EBADENGINE   required: { node: '^14.15.0 || ^16.10.0 || >=18.0.0' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'jest-diff@29.3.1',
npm WARN EBADENGINE   required: { node: '^14.15.0 || ^16.10.0 || >=18.0.0' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'pretty-format@29.3.1',
npm WARN EBADENGINE   required: { node: '^14.15.0 || ^16.10.0 || >=18.0.0' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'jest-get-type@29.2.0',
npm WARN EBADENGINE   required: { node: '^14.15.0 || ^16.10.0 || >=18.0.0' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'jest-matcher-utils@29.3.1',
npm WARN EBADENGINE   required: { node: '^14.15.0 || ^16.10.0 || >=18.0.0' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'pretty-format@29.3.1',
npm WARN EBADENGINE   required: { node: '^14.15.0 || ^16.10.0 || >=18.0.0' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'jest-message-util@29.3.1',
npm WARN EBADENGINE   required: { node: '^14.15.0 || ^16.10.0 || >=18.0.0' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'pretty-format@29.3.1',
npm WARN EBADENGINE   required: { node: '^14.15.0 || ^16.10.0 || >=18.0.0' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'jest-util@29.3.1',
npm WARN EBADENGINE   required: { node: '^14.15.0 || ^16.10.0 || >=18.0.0' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'react-app-polyfill@3.0.0',
npm WARN EBADENGINE   required: { node: '>=14' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'react-dev-utils@12.0.1',
npm WARN EBADENGINE   required: { node: '>=14' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'react-scripts@5.0.1',
npm WARN EBADENGINE   required: { node: '>=14.0.0' },
npm WARN EBADENGINE   current: { node: 'v12.22.9', npm: '8.5.1' }
npm WARN EBADENGINE }

up to date, audited 1484 packages in 3s

291 packages are looking for funding
  run `npm fund` for details

6 high severity vulnerabilities

To address all issues (including breaking changes), run:
  npm audit fix --force

Run `npm audit` for details
```

**Reply by jas3333** [reply]
If you're not getting a response it's most likely a API key issue. Did you create a file named `.env` in the client directory and create a variable called `REACT_APP_OPENAI_KEY=yourkey`? It has to be named `.env` and nothing else.

**Reply by jose97ct** [reply]
Exactly like that. Same name, same content with no spaces or symbols. Should i try on a Windows machine instead of Linux?

**Reply by jas3333** [reply]
I run it on a Linux machine. What error do you get when you open up the dev tools in the browser?

**Comment by jas3333** [comment]
From those errors, looks like you are using a really old version of Node. Which Linux distro are you using? Your version of Node should be above 14 looks like you're on 12.22.9

**Reply by jose97ct** [reply]
Hi, I've tried updating node via `apt update`, `npm update -g` and `npm install npm@latest` with no results. Is there any other way to force upgrading node or uninstalling and installing a newer version? I've actually never used npm before

Btw my browser's console outputs: `Failed to load resource: the server responded with a status of 429 ()`

**Reply by jose97ct** [reply]
Okay, I've just found how to update Node and npm, now i have `v12.22.9` and `9.2.0` respectively.

Also I've found this message in the browser's console when I hit enter:
`AxiosError {message: 'Request failed with status code 429', name: 'AxiosError', code: 'ERR_BAD_REQUEST', config: {…}, request: XMLHttpRequest, …}`

Tried again reinstalling everything after all the updates, but same result

**Reply by jas3333** [reply]
A 429 error means too many requests. Still looks like you're using an old version of node. Try installing from here 
[Node Website]https://nodejs.org/en/)

**Reply by jas3333** [reply]
Ok did some experimenting on my Ubuntu server and it was stuck on same version you have. Try this: 
```
sudo npm cache clean -f
sudo npm install -g n
sudo n stable
```

After doing all of that close your terminal and reopen it. `node -v` it should be that latest stable version.

**Reply by jose97ct** [reply]
I've now managed to update nodejs and npm, I'm currently running v18.12.1 and 8.19.2. Removed the entire GPT-Helper folder and started from the beginning. 
Now I'm not getting any warnings during `npm install` except of the `6 high severity vulnerabilities` message at the end of the install. Tried again with the same error 429 in the browser's console.
This is my `java --version` output:
```
openjdk 18.0.2-ea 2022-07-19
OpenJDK Runtime Environment (build 18.0.2-ea+9-Ubuntu-222.04)
OpenJDK 64-Bit Server VM (build 18.0.2-ea+9-Ubuntu-222.04, mixed mode, sharing)
```

And my browser's console output: 
```
xhr.js:247          POST https://api.openai.com/v1/completions 429
dispatchXhrRequest @ xhr.js:247
xhr @ xhr.js:49
dispatchRequest @ dispatchRequest.js:51
request @ Axios.js:142
httpMethod @ Axios.js:181
wrap @ bind.js:5
onSubmit @ Home.js:70
handleSubmit @ Prompt.js:16
callCallback @ react-dom.development.js:4164
invokeGuardedCallbackDev @ react-dom.development.js:4213
invokeGuardedCallback @ react-dom.development.js:4277
invokeGuardedCallbackAndCatchFirstError @ react-dom.development.js:4291
executeDispatch @ react-dom.development.js:9041
processDispatchQueueItemsInOrder @ react-dom.development.js:9073
processDispatchQueue @ react-dom.development.js:9086
dispatchEventsForPlugins @ react-dom.development.js:9097
(anonymous) @ react-dom.development.js:9288
batchedUpdates$1 @ react-dom.development.js:26140
batchedUpdates @ react-dom.development.js:3991
dispatchEventForPluginEventSystem @ react-dom.development.js:9287
dispatchEventWithEnableCapturePhaseSelectiveHydrationWithoutDiscreteEventReplay @ react-dom.development.js:6465
dispatchEvent @ react-dom.development.js:6457
dispatchDiscreteEvent @ react-dom.development.js:6430
Home.js:85 AxiosError {message: 'Request failed with status code 429', name: 'AxiosError', code: 'ERR_BAD_REQUEST', config: {…}, request: XMLHttpRequest, …}
```
It seems to be something related to React, but I'm not really sure as I'm learning on the fly. 

Thank you again for your help, I feel like we are just one step away from achieving this!

**Comment by jose97ct** [comment]
Wait, it seems to be some warnings left in the installation process:
```
npm WARN deprecated stable@0.1.8: Modern JS already guarantees Array#sort() is a stable sort, so this library is deprecated. See the compatibility table on MDN: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#browser_compatibility
npm WARN deprecated rollup-plugin-terser@7.0.2: This package has been deprecated and is no longer maintained. Please use @rollup/plugin-terser
npm WARN deprecated w3c-hr-time@1.0.2: Use your platform's native performance.now() and performance.timeOrigin.
npm WARN deprecated sourcemap-codec@1.4.8: Please use @jridgewell/sourcemap-codec instead
npm WARN deprecated svgo@1.3.2: This SVGO version is no longer supported. Upgrade to v2.x.x.
```
Hope all this info helps, and sorry for my inconvenience

**Reply by jas3333** [reply]
Yeah I get those errors too, I just ignore them. The issue is the 429 response you are getting which means your key is rate limited. The only reason I can see you getting that error is if you spammed the enter key as it would be trying to make too many request. You should see a loading indicator on the right side of the chat box if your request was sent out.

**Reply by jose97ct** [reply]
Idk, but I'm still having the same quota error on new generated API keys, fresh GPT-Helper installation and without spamming enter. The loading indicator works for less than one second while the console displays the 429 error. I will try in another fresh installed lubuntu virtual machine, as it is the OS I'm using in my physical server

**Reply by jose97ct** [reply]
I've just tried in a new Virtual Machine and the key of another OpenAI account with same result. Console displays POST in red, and error 429. I have opened the server GUI to Internet in case you want to test it by yourself. I can DM you the link if you want to try. 
Thank you as always

**Reply by jas3333** [reply]
Sure you can send me the link. I really don't understand why you are getting rate limited.

**Reply by jose97ct** [reply]
It seems that Github removed private messaging some time ago, so here you have the link `http://maincraxavales.ddns.net:3000`. It's a test machine I use for hosting a Minecraft server, so be free to do whatever you want

**Reply by jas3333** [reply]
Did you host this publicly to start? Because it sounds like it got swarmed by bots or something and has been rate limited. Looks like you'll have to contact OpenAI. I would suggest taking this down immediately though, as it's not designed to be on the web. Your API key is fully exposed.

**Reply by jose97ct** [reply]
I just did it for our research, but I usually don't have it open to internet because of obvious reasons. I will contact OpenAI for the quota thing because it's the last thing to be fixed for it to work. I really appreciate your efforts trying to help me these days, thank you!

**Reply by jas3333** [reply]
You're welcome. Hopefully it can get sorted out.
